File: 이것이 코딩테스트다/07.최단 경로/4.미래도시.js
const texts = [
  [5, 7],
  [1, 2],
  [1, 3],
  [1, 4],
  [2, 4],
  [3, 4],
  [3, 5],
  [4, 5],
  [4, 5],
];

const [n, m] = texts.shift();
const [x, k] = texts.pop();
const INF = Number(Infinity);
let graph = Array.from({ length: n + 1 }, () => Array(n + 1).fill(INF));

for (const [a, b] of texts) {
  graph[a][b] = 1;
  graph[b][a] = 1;
}

for (let a = 1; a < n + 1; a++) {
  for (let b = 1; b < n + 1; b++) {
    if (a === b) {
      graph[a][b] = 0;
    }
  }
}
for (let k = 1; k < n + 1; k++) {
  for (let i = 1; i < n + 1; i++) {
    for (let j = 1; j < n + 1; j++) {
      graph[i][j] = Math.min(graph[i][j], graph[i][k] + graph[k][j]);
    }
  }
}
let result = graph[1][k] + graph[k][x];

if (result >= INF) {
  console.log(-1);
} else {
  console.log(result);
}

File: 이것이 코딩테스트다/07.최단 경로/4.미래도시.py
# 1번 회사에 A가 위치 X번 회사에 방문해 물건 판매
# 양방향 이동이 가능, 이동 1시간 소요
# 1번 -> k번 -> X번

# 회사 개수, 경로 개수
n, m = map(int, input().split())

INF = int(1e9)

graph = [[INF] * (n+1) for _ in range(n+1)]

for _ in range(m):
  a, b = map(int, input().split())
  graph[a][b] = 1
  graph[b][a] = 1
x, k = map(int, input().split())
for i in range(1, n+1):
  for j in range(1, n+1):
    if i == j:
      graph[i][j] = 0


for k in range(1, n+1):
  for i in range(1, n+1):
    for j in range(1, n+1):
      graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j])

result = graph[1][k] + graph[k][x]
if result >= INF:
  print(-1)
else:
  print(result)